File: scrape.py
# scrape.py
import config
import pymysql.cursors
import googlemaps
import requests
from BeautifulSoup import BeautifulSoup

#--Configuration--
gmaps = googlemaps.Client(key = config.GMAPS_API_KEY)
mysql = pymysql.connect(host = config.MYSQL_HOST,
                             user = config.MYSQL_USERNAME ,
                             password = config.MYSQL_PASSWORD,
                             db = config.MYSQL_DATABASE,
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

#--Address to GPS Coordinates--
def addressToGPS(address):
    # 'gmaps.geocode()' returns data in JSON format.
    locationData = gmaps.geocode(address)
    location     = locationData[0]['geometry']['location']
    latitude     = location['lat']
    longitude    = location['lng']
    return (latitude, longitude)

#--Access and Parse Webpage-- 
# Access time for url and html parsing is slow. One possible cause is advertisements on the foodnetwork webpages.
def makeSoup(url):
    response = requests.get(url)
    html     = response.content
    return     BeautifulSoup(html)

#--Scrape Restaurant Data--
def getRestaurantData(soup):
    # Search time for 'candidateTags' is slow.
    # One possible solution is to search for candidates within their parent tag as opposed to the entire html document.
    candidateTags  = soup.findAll('div', attrs = {"class": "m-MediaBlock o-Capsule__m-MediaBlock"})
    # A list to hold useful restaurant data.
    restaurantData = []
    # Assign an index to compare restaurants by the order in which they appear. Restaurant names are not distinct.
    index = 1

    for tag in candidateTags:
        # Only candidateTags with 'h3' header tags hold restaurant data.
        if tag.find('h3'):
            header   = tag.find('h3', attrs = {"class": "m-MediaBlock__a-Headline"})
            name     = header.span.text
            # The url string comes formatted: "//www...". It is formatted to "www...".
            url      = header.a.get('href').replace("//",'')
            # The address string comes formatted: "...[State]&nbsp;[Zipcode]". It is formatted to "...[State] [Zipcode]".
            address  = tag.find('div', attrs = {"class": "m-Info__a-Address"}).text.replace('&nbsp;',' ')
            # Location of the restaurant given in GPS coordinates
            location = addressToGPS(address)
            restaurantData.append((index, name, location, address, url))
            index += 1

    return restaurantData

# --Get All Restaurant Data--
# Create Page URLs
# Each page url is composed of the 'baseURL' followed by an unsigned, sequential integer index value -- currently 1 - 66 inclusive.
baseURL = "http://www.foodnetwork.com/restaurants/shows/diners-drive-ins-and-dives/a-z/p/"
pages   = []
# Hard-coding the maximum page number is a bad idea.
# The number of pages will change over time. Find another method.
for i in range(1, 2):
    pages.append(baseURL + str(i))

# Access, parse, and scrape each page.
restaurantData = []
for page in pages:
    soup            = makeSoup(page)
    restaurantData += getRestaurantData(soup)

#--Store Restaurant Data in Database--
for restaurant in restaurantData:
    identifier = str(restaurant[0])
    name       = restaurant[1]
    lat        = str(restaurant[2][0])
    lng        = str(restaurant[2][1])
    address    = restaurant[3]
    url        = restaurant[4]
    cursor     = mysql.cursor()
    addInfo    = "(id, name, latitude, longitude, address, url) VALUES (" + identifier + ", \"" + name + "\", " + lat + ", " + lng + ", \"" + address + "\", \"" + url + "\")"
    # updateInfo modifies all fields except for 'id'
    updateInfo = "name=\"" + name + "\", latitude=" + lat + ", longitude=" + lng + ", address=\"" + address + "\", url=\"" + url  + "\""
    # If the restaurant is not already in the table (i.e. its id is not present in the table) then add it to the table.
    # If the restaurant exists, update its information.
    query  = "INSERT INTO restaurants " + addInfo + "ON DUPLICATE KEY UPDATE " + updateInfo
    cursor.execute(query)
    mysql.commit()
    cursor.close()

File: scrape.php

<!-- scrape.php - 03/24/2018-->
<?php
    // config.php defines SQL information and API keys.
    include "config.php";
    // A web scraping library.
    include "simple_html_dom.php";

    function getLocation($address){
        // Format the address to be inserted into an api request url.
        $formatted_address = str_replace(" ","+",str_replace(",","", trim($address)));
        // Build the gmaps geocode http request url.
        $url = "https://maps.googleapis.com/maps/api/geocode/json?address=" . $formatted_address . "&key=" . GMAPS_KEY;
        // Retrieves jSON object from gmaps geocode api and decodes it into an array.
        $response = json_decode(file_get_contents($url), true);
        // Retrieves location array with children "lat" and "lng".
        $location = $response["results"][0]["geometry"]["location"];

        return $location;
    }
    
    function scrape(){
        // Open database connection.
        $db = mysqli_connect(SQL_HOST, SQL_USER, SQL_PASSWORD, SQL_DATABASE)
        or die("Error connecting to MySQL server.");

        $page_max = 2;
        $id       = 1;
        for($page = 1; $page <= $page_max; $page++){
            $fn_url = "http://www.foodnetwork.com/restaurants/shows/diners-drive-ins-and-dives/a-z/p/" . $page;
            $html   = file_get_html($fn_url);

            foreach ($html->find("div[class=m-MediaBlock o-Capsule__m-MediaBlock]") as $candidate){
                // Candidate is a restaurant header if it has a child h3 tag.
                if($candidate->find("h3")){
                    // The header is the first h3 tag of the candidate div.
                    $header   = $candidate -> find("h3[class=m-MediaBlock__a-Headline]", 0);
                    $name     = $header -> find("span", 0) -> plaintext;
                    $url      = str_replace("//", "", $header -> find("a", 0) -> href);
                    $address  = trim(str_replace("&nbsp;", " ", $candidate -> find("div[class=m-Info__a-Address]",0) -> plaintext));
                    $location = getLocation($address);
                    $lat      = $location["lat"];
                    $lng      = $location["lng"];

                    $addInfo    = "(id, name, latitude, longitude, address, url) VALUES (" . $id . ", \"" . $name . "\", " . $lat . ", " . $lng . ", \"" . $address . "\", \"" . $url . "\")";
                    $updateInfo = "name=\"" . $name . "\", latitude=" . $lat . ", longitude=" . $lng . ", address=\"" . $address . "\", url=\"" . $url  . "\"";
                    $query      = "INSERT INTO " . SQL_TABLE . " " . $addInfo . " ON DUPLICATE KEY UPDATE " . $updateInfo;

                    mysqli_query($db, $query);
                    $id += 1;
                }
            }
        }
        
        // Commit database changes.
        mysqli_commit($db);
        // Close database connection.
        mysqli_close($db);
    }

    
?>
